File: programmers/pg_lv2_게임맵최단거리.js
function solution(maps) {
  const n = maps.length;
  const m = maps[0].length;
  const graph = [new Array(m + 2).fill(0)];
  maps.forEach((ele) => {
    graph.push([0, ...ele, 0]);
  });
  graph.push(new Array(m + 2).fill(0));

  const dy = [-1, 1, 0, 0];
  const dx = [0, 0, -1, 1];
  const queue = [];
  let head = 0;

  const bfs = (i, j) => {
    queue.push([i, j, 1]);
    graph[i][j] = 0;

    while (queue.length > head) {
      const [y, x, count] = queue[head++];

      for (d = 0; d < 4; d++) {
        const ny = y + dy[d];
        const nx = x + dx[d];

        if (ny == n && nx == m) return count + 1;

        if (graph[ny][nx] == 1) {
          queue.push([ny, nx, count + 1]);
          graph[ny][nx] = 0;
        }
      }
    }
    return -1;
  };

  return bfs(1, 1);
}

File: python/programmers/pg_lv2_게임맵최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    dy = [-1, 1, 0, 0]
    dx = [0, 0, -1, 1]
    visited = [[False for _ in range(m)] for _ in range(n)]
    
    def bfs(i, j, c):
        queue = deque([(i, j, c)])
        visited[i][j] = True
        
        while queue:
            y, x, count = queue.popleft()
            
            for d in range(4):
                ny = y + dy[d]
                nx = x + dx[d]
                
                if ny == n - 1 and nx == m - 1: return count + 1
                if ny < 0 or nx < 0 or ny >= n or nx >= m or maps[ny][nx] == 0: continue
                if not visited[ny][nx]:
                    visited[ny][nx] = True
                    queue.append((ny, nx, count + 1))
        return -1
    
    return bfs(0, 0, 1)